pkg/validator/validators.go

```diff
@@ -266,3 +268,34 @@ func (kc *policyValidator) upstreamTrafficSettingValidator(req *admissionv1.Admi
 
 	return nil, nil
 }
+
+func meshRootCertificateValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	switch req.Operation {
+	case admissionv1.Create:
+		newMRC := &configv1alpha2.MeshRootCertificate{}
+		if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(newMRC); err != nil {
+			return nil, err
+		}
+
+		if newMRC.Spec.TrustDomain == "" {
+			return nil, fmt.Errorf("trustDomain must be non empty for MRC %s/%s", newMRC.GetNamespace(), newMRC.GetName())
+		}
+	case admissionv1.Update:
+		newMRC, oldMRC := &configv1alpha2.MeshRootCertificate{}, &configv1alpha2.MeshRootCertificate{}
+		if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(newMRC); err != nil {
+			return nil, err
+		}
+		if err := json.NewDecoder(bytes.NewBuffer(req.OldObject.Raw)).Decode(oldMRC); err != nil {
+			return nil, err
+		}
+
+		if !reflect.DeepEqual(oldMRC.Spec.Provider, newMRC.Spec.Provider) {
+			return nil, fmt.Errorf("cannot update certificate provider settings for MRC %s/%s. Create a new MRC and initiate root certificate rotation to update the provider", newMRC.GetNamespace(), newMRC.GetName())
+		}
+
+		if oldMRC.Spec.TrustDomain != newMRC.Spec.TrustDomain {
+			return nil, fmt.Errorf("cannot update trust domain for MRC %s/%s. Create a new MRC and initiate root certificate rotation to update the trust domain", newMRC.GetNamespace(), newMRC.GetName())
+		}
+	}
```

Comment:
have you tested the semantics of preventing a deletion of an active MRC with uninstalling the cluster?

Reply:
I have not, will make a note of this in the related issue if it isn't already there.